pkg/controllers/overrider/common.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package overrider features controllers to reconcile the override objects.
+package overrider
+
+import (
+	"context"
+	"sort"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	placementv1alpha1 "go.goms.io/fleet/apis/placement/v1alpha1"
+	placementv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+	"go.goms.io/fleet/pkg/utils/controller"
+	"go.goms.io/fleet/pkg/utils/labels"
+)
+
+// Reconciler reconciles a clusterResourceOverride object.
+type Reconciler struct {
+	// Client is used to update objects which goes to the api server directly.
+	client.Client
+}
+
+// handleOverrideDeleting handles the delete event of an override object. We need to delete all the related override Snapshot.
+func (r *Reconciler) handleOverrideDeleting(ctx context.Context, overrideSnapshotObj, parentOverrideObj client.Object) error {
+	croObjRef := klog.KObj(parentOverrideObj)
+	if !controllerutil.ContainsFinalizer(parentOverrideObj, placementv1alpha1.OverrideFinalizer) {
+		klog.V(2).InfoS("No need to do anything for the deleting override without a finalizer", "clusterResourceOverride", croObjRef)
+		return nil
+	}
+	// delete all the associated snapshots
+	if err := r.Client.DeleteAllOf(ctx, overrideSnapshotObj, client.MatchingLabels{placementv1alpha1.OverrideTrackingLabel: parentOverrideObj.GetName()}); err != nil {
+		klog.ErrorS(err, "Failed to delete all associated overrideSnapshot", "clusterResourceOverride", croObjRef)
+		return err
```

Comment:
add the error wrapper?

what if the associated snapshot is 0, will delete fail?

Reply:
Add the NewAPIServerError error wrapper?